File: netlify/functions/gov-vote-notify.ts
import { Config } from "@netlify/functions";
import { getStore } from "@netlify/blobs";
import { bech32 } from "bech32";

interface Vote {
  voter: string;
  options: Array<{ option: string; weight: string }>;
}

interface Validator {
  operator_address: string;
  tokens: string;
  description: {
    moniker: string;
  };
}

interface ValidatorInfo {
  moniker: string;
  tokens: number;
  operatorAddress: string;
  votingPowerPct: number;
}

interface SeenVotes {
  [voter: string]: {
    option: string;
    timestamp: string;
  };
}

// Scheduled function configuration
export const config: Config = {
  schedule: "* * * * *", // Every minute
};

export default async function handler() {
  const lcdUrl = process.env.COSMOS_LCD_URL || "https://cosmos-lcd.publicnode.com";
  const proposalIds = (process.env.PROPOSAL_IDS || "").split(",").filter(Boolean);
  const slackWebhook = process.env.SLACK_WEBHOOK_URL;
  const errorWebhook = process.env.SLACK_ERROR_WEBHOOK_URL;

  if (proposalIds.length === 0) {
    return new Response("PROPOSAL_IDS not configured", { status: 400 });
  }

  if (!slackWebhook) {
    return new Response("SLACK_WEBHOOK_URL not configured", { status: 400 });
  }

  const store = getStore("gov-votes");

  try {
    for (const proposalId of proposalIds) {
      await processProposal(proposalId.trim(), lcdUrl, store, slackWebhook);
    }
    return new Response("OK", { status: 200 });
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    if (errorWebhook) {
      await postErrorToSlack(errorWebhook, errorMessage);
    }
    throw error;
  }
}

async function processProposal(
  proposalId: string,
  lcdUrl: string,
  store: ReturnType<typeof getStore>,
  slackWebhook: string
): Promise<number> {
  // Get all votes for the proposal
  const votes = await fetchAllVotes(lcdUrl, proposalId);

  // Get validators for moniker + voting power lookup
  const validators = await fetchValidators(lcdUrl);

  // Build mapping from account addresses to validator info
  const validatorByAccount = buildAccountToValidatorMapping(validators);

  // Load previously seen votes with strong consistency to avoid duplicates
  const seenKey = `seen_votes_${proposalId}`;
  let seenVotes: SeenVotes = {};
  try {
    const seenData = await store.get(seenKey, { type: "json", consistency: "strong" }) as SeenVotes | null;
    if (seenData) {
      seenVotes = seenData;
    }
  } catch {
    // No existing data, start fresh
  }

  // Calculate total staked tokens for percentage calculations
  let totalStakedTokens = 0;
  for (const validator of validators.values()) {
    totalStakedTokens += validator.tokens;
  }

  const newValidatorVotes: Array<{ vote: Vote; validator: ValidatorInfo }> = [];
  const newWhaleVotes: Array<{ vote: Vote; votingPowerPct: number }> = [];

  for (const vote of votes) {
    if (!(vote.voter in seenVotes)) {
      // Mark as seen regardless of whether it's a validator
      seenVotes[vote.voter] = {
        option: vote.options?.[0]?.option || "UNKNOWN",
        timestamp: new Date().toISOString(),
      };

      // Check if it's a validator vote
      const validator = validatorByAccount.get(vote.voter);
      if (validator) {
        newValidatorVotes.push({ vote, validator });
      } else {
        // Check if it's a whale delegator (>0.05% VP)
        const delegatorVP = await fetchDelegatorVotingPower(lcdUrl, vote.voter, totalStakedTokens);
        if (delegatorVP >= 0.05) {
          newWhaleVotes.push({ vote, votingPowerPct: delegatorVP });
        }
      }
    }
  }

  // Notify for each new validator vote
  for (const { vote, validator } of newValidatorVotes) {
    await notifyVote(vote, proposalId, validator, slackWebhook);
  }

  // Notify for each new whale vote
  for (const { vote, votingPowerPct } of newWhaleVotes) {
    await notifyWhaleVote(vote, proposalId, votingPowerPct, slackWebhook);
  }

  // Persist updated seen votes
  await store.setJSON(seenKey, seenVotes);

  return newValidatorVotes.length;
}

async function fetchAllVotes(lcdUrl: string, proposalId: string): Promise<Vote[]> {
  const votes: Vote[] = [];
  let paginationKey: string | null = null;

  while (true) {
    const url = new URL(`${lcdUrl}/cosmos/gov/v1/proposals/${proposalId}/votes`);
    url.searchParams.set("pagination.limit", "100");
    if (paginationKey) {
      url.searchParams.set("pagination.key", paginationKey);
    }

    const resp = await fetch(url.toString(), {
      signal: AbortSignal.timeout(30000),
    });

    if (!resp.ok) {
      throw new Error(`Failed to fetch votes: ${resp.status} ${resp.statusText}`);
    }

    const data = await resp.json();
    votes.push(...(data.votes || []));
    paginationKey = data.pagination?.next_key || null;

    if (!paginationKey) {
      break;
    }
  }

  return votes;
}

async function fetchValidators(
  lcdUrl: string
): Promise<Map<string, ValidatorInfo>> {
  const validators = new Map<string, ValidatorInfo>();
  let paginationKey: string | null = null;
  let totalPower = 0;

  // First pass: collect all validators and total power
  const validatorList: Array<{ address: string; info: ValidatorInfo }> = [];

  while (true) {
    const url = new URL(`${lcdUrl}/cosmos/staking/v1beta1/validators`);
    url.searchParams.set("pagination.limit", "100");
    url.searchParams.set("status", "BOND_STATUS_BONDED");
    if (paginationKey) {
      url.searchParams.set("pagination.key", paginationKey);
    }

    const resp = await fetch(url.toString(), {
      signal: AbortSignal.timeout(30000),
    });

    if (!resp.ok) {
      throw new Error(`Failed to fetch validators: ${resp.status} ${resp.statusText}`);
    }

    const data = await resp.json();

    for (const v of data.validators || []) {
      const tokens = parseInt(v.tokens, 10);
      totalPower += tokens;
      validatorList.push({
        address: v.operator_address,
        info: {
          moniker: v.description.moniker,
          tokens,
          operatorAddress: v.operator_address,
          votingPowerPct: 0, // Calculate after we have total
        },
      });
    }

    paginationKey = data.pagination?.next_key || null;
    if (!paginationKey) {
      break;
    }
  }

  // Second pass: calculate voting power percentages
  for (const { address, info } of validatorList) {
    info.votingPowerPct = totalPower > 0 ? (info.tokens / totalPower) * 100 : 0;
    validators.set(address, info);
  }

  return validators;
}

function operatorToAccount(operatorAddress: string): string | null {
  try {
    const decoded = bech32.decode(operatorAddress);
    if (decoded.prefix === "cosmosvaloper") {
      return bech32.encode("cosmos", decoded.words);
    }
  } catch {
    // Invalid address
  }
  return null;
}

function buildAccountToValidatorMapping(
  validators: Map<string, ValidatorInfo>
): Map<string, ValidatorInfo> {
  const mapping = new Map<string, ValidatorInfo>();

  for (const [operatorAddress, info] of validators) {
    const accountAddress = operatorToAccount(operatorAddress);
    if (accountAddress) {
      mapping.set(accountAddress, info);
    }
  }

  return mapping;
}

async function notifyVote(
  vote: Vote,
  proposalId: string,
  validator: ValidatorInfo,
  slackWebhook: string
): Promise<void> {
  const rawOption = vote.options?.[0]?.option || "UNKNOWN";
  const option = formatVoteOption(rawOption);
  const emoji = getVoteEmoji(rawOption);
  const timestamp = new Date().toISOString().replace("T", " ").slice(0, 16) + " UTC";

  const { moniker, votingPowerPct } = validator;
  const message = `${emoji} ${moniker} (${votingPowerPct.toFixed(1)}% VP) votes ${option} on proposal ${proposalId} at ${timestamp}`;

  const resp = await fetch(slackWebhook, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ text: message }),
    signal: AbortSignal.timeout(10000),
  });

  if (!resp.ok) {
    throw new Error(`Failed to post to Slack: ${resp.status} ${resp.statusText}`);
  }
}

function formatVoteOption(option: string): string {
  const mapping: Record<string, string> = {
    VOTE_OPTION_YES: "Yes",
    VOTE_OPTION_NO: "No",
    VOTE_OPTION_ABSTAIN: "Abstain",
    VOTE_OPTION_NO_WITH_VETO: "NoWithVeto",
  };
  return mapping[option] || option;
}

function getVoteEmoji(option: string): string {
  switch (option) {
    case "VOTE_OPTION_YES":
      return "\u2705\u2705\u2705"; // Green checkmarks
    case "VOTE_OPTION_NO":
      return "\u{1F6A8}\u{1F6A8}\u{1F6A8}\u26A0\uFE0F\u26A0\uFE0F\u26A0\uFE0F\u{1F6A8}\u{1F6A8}\u{1F6A8}"; // Red sirens and warnings
    case "VOTE_OPTION_ABSTAIN":
      return "\u{1F7E4}\u{1F7E4}\u{1F7E4}"; // Gray circles
    case "VOTE_OPTION_NO_WITH_VETO":
      return "\u{1F6A8}\u{1F6A8}\u{1F6A8}\u26A0\uFE0F\u26A0\uFE0F\u26A0\uFE0F\u274C\u274C\u274C"; // Sirens, warnings, and X marks
    default:
      return "\u2753"; // Question mark
  }
}

async function fetchDelegatorVotingPower(
  lcdUrl: string,
  delegatorAddress: string,
  totalStakedTokens: number
): Promise<number> {
  try {
    const url = new URL(`${lcdUrl}/cosmos/staking/v1beta1/delegations/${delegatorAddress}`);
    const resp = await fetch(url.toString(), {
      signal: AbortSignal.timeout(10000),
    });

    if (!resp.ok) {
      return 0;
    }

    const data = await resp.json();
    const delegations = data.delegation_responses || [];

    let totalDelegated = 0;
    for (const delegation of delegations) {
      totalDelegated += parseInt(delegation.balance?.amount || "0", 10);
    }

    if (totalStakedTokens === 0) return 0;
    return (totalDelegated / totalStakedTokens) * 100;
  } catch {
    return 0;
  }
}

async function notifyWhaleVote(
  vote: Vote,
  proposalId: string,
  votingPowerPct: number,
  slackWebhook: string
): Promise<void> {
  const rawOption = vote.options?.[0]?.option || "UNKNOWN";
  const option = formatVoteOption(rawOption);
  const voteEmoji = getVoteEmoji(rawOption);
  const timestamp = new Date().toISOString().replace("T", " ").slice(0, 16) + " UTC";

  const whaleEmoji = "\u{1F433}\u{1F433}\u{1F433}"; // Whale emojis
  const shortVoter = vote.voter.length > 20
    ? `${vote.voter.slice(0, 12)}...${vote.voter.slice(-6)}`
    : vote.voter;

  const message = `${whaleEmoji} *WHALE ALERT* ${whaleEmoji}\n${voteEmoji} ${shortVoter} (${votingPowerPct.toFixed(2)}% VP) votes ${option} on proposal ${proposalId} at ${timestamp}`;

  const resp = await fetch(slackWebhook, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ text: message }),
    signal: AbortSignal.timeout(10000),
  });

  if (!resp.ok) {
    throw new Error(`Failed to post to Slack: ${resp.status} ${resp.statusText}`);
  }
}

async function postErrorToSlack(webhook: string, errorMessage: string): Promise<void> {
  try {
    await fetch(webhook, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        text: `:warning: Gov notification bot error: ${errorMessage}`,
      }),
      signal: AbortSignal.timeout(10000),
    });
  } catch {
    // Don't throw if error notification fails
  }
}

File: netlify/functions/gov-vote-notify/gov-vote-notify.py
"""
Cosmos Hub Governance Vote Notification Bot

A serverless function that monitors Cosmos Hub validator votes on specified
governance proposals and posts notifications to Slack.
"""

import os
import json
import requests
from datetime import datetime, timezone
import bech32


def handler(event, context):
    """Scheduled function that runs every 1 minute."""
    lcd_url = os.environ.get("COSMOS_LCD_URL", "https://cosmos-lcd.publicnode.com")
    proposal_ids = os.environ.get("PROPOSAL_IDS", "").split(",")
    slack_webhook = os.environ.get("SLACK_WEBHOOK_URL")
    error_webhook = os.environ.get("SLACK_ERROR_WEBHOOK_URL")

    if not proposal_ids or not proposal_ids[0]:
        return {"statusCode": 400, "body": "PROPOSAL_IDS not configured"}

    if not slack_webhook:
        return {"statusCode": 400, "body": "SLACK_WEBHOOK_URL not configured"}

    # Use in-memory store for local testing, Netlify Blobs in production
    store = get_store()

    try:
        for proposal_id in proposal_ids:
            proposal_id = proposal_id.strip()
            if proposal_id:
                process_proposal(proposal_id, lcd_url, store, slack_webhook)
        return {"statusCode": 200, "body": "OK"}
    except Exception as e:
        if error_webhook:
            post_error_to_slack(error_webhook, str(e))
        raise


def get_store():
    """Get appropriate storage backend based on environment."""
    try:
        from netlify_blobs import get_store as get_netlify_store
        return NetlifyBlobStore(get_netlify_store("gov-votes"))
    except ImportError:
        # Fall back to file-based storage for local testing
        return FileStore()


class NetlifyBlobStore:
    """Wrapper for Netlify Blobs storage."""

    def __init__(self, store):
        self.store = store

    def get(self, key):
        return self.store.get(key)

    def set(self, key, value):
        self.store.set(key, value)


class FileStore:
    """File-based storage for local testing."""

    def __init__(self, storage_dir=None):
        if storage_dir is None:
            storage_dir = os.path.join(os.path.dirname(__file__), ".store")
        self.storage_dir = storage_dir
        os.makedirs(self.storage_dir, exist_ok=True)

    def _get_path(self, key):
        # Sanitize key for filesystem
        safe_key = key.replace("/", "_").replace("\\", "_")
        return os.path.join(self.storage_dir, f"{safe_key}.json")

    def get(self, key):
        path = self._get_path(key)
        if os.path.exists(path):
            with open(path, "r") as f:
                return f.read()
        return None

    def set(self, key, value):
        path = self._get_path(key)
        with open(path, "w") as f:
            f.write(value)


def process_proposal(proposal_id, lcd_url, store, slack_webhook):
    """Fetch votes for a proposal and notify on new ones."""
    # Get all votes for the proposal
    votes = fetch_all_votes(lcd_url, proposal_id)

    # Get validators for moniker + voting power lookup
    validators = fetch_validators(lcd_url)

    # Build mapping from account addresses to validator info
    validator_by_account = build_account_to_validator_mapping(validators)

    # Load previously seen votes
    seen_key = f"seen_votes_{proposal_id}"
    seen_data = store.get(seen_key)
    seen_votes = json.loads(seen_data) if seen_data else {}

    new_votes = []
    for vote in votes:
        voter = vote["voter"]
        if voter not in seen_votes:
            new_votes.append(vote)
            seen_votes[voter] = {
                "option": vote["options"][0]["option"] if vote.get("options") else "UNKNOWN",
                "timestamp": datetime.now(timezone.utc).isoformat()
            }

    # Notify for each new vote
    for vote in new_votes:
        notify_vote(vote, proposal_id, validator_by_account, slack_webhook)

    # Persist updated seen votes
    store.set(seen_key, json.dumps(seen_votes))

    return len(new_votes)


def fetch_all_votes(lcd_url, proposal_id):
    """Fetch all votes with pagination."""
    votes = []
    pagination_key = None

    while True:
        url = f"{lcd_url}/cosmos/gov/v1/proposals/{proposal_id}/votes"
        params = {"pagination.limit": "100"}
        if pagination_key:
            params["pagination.key"] = pagination_key

        resp = requests.get(url, params=params, timeout=30)
        resp.raise_for_status()
        data = resp.json()

        votes.extend(data.get("votes", []))
        pagination_key = data.get("pagination", {}).get("next_key")

        if not pagination_key:
            break

    return votes


def fetch_validators(lcd_url):
    """Fetch all bonded validators with their monikers and voting power."""
    validators = {}
    pagination_key = None
    total_power = 0

    while True:
        url = f"{lcd_url}/cosmos/staking/v1beta1/validators"
        params = {"pagination.limit": "100", "status": "BOND_STATUS_BONDED"}
        if pagination_key:
            params["pagination.key"] = pagination_key

        resp = requests.get(url, params=params, timeout=30)
        resp.raise_for_status()
        data = resp.json()

        for v in data.get("validators", []):
            tokens = int(v["tokens"])
            total_power += tokens
            validators[v["operator_address"]] = {
                "moniker": v["description"]["moniker"],
                "tokens": tokens,
                "operator_address": v["operator_address"]
            }

        pagination_key = data.get("pagination", {}).get("next_key")
        if not pagination_key:
            break

    # Calculate voting power percentages
    for v in validators.values():
        v["voting_power_pct"] = (v["tokens"] / total_power) * 100 if total_power > 0 else 0

    return validators


def operator_to_account(operator_address):
    """Convert cosmosvaloper1... to cosmos1... address."""
    try:
        hrp, data = bech32.bech32_decode(operator_address)
        if hrp == "cosmosvaloper":
            return bech32.bech32_encode("cosmos", data)
    except Exception:
        pass
    return None


def build_account_to_validator_mapping(validators):
    """Build a mapping from account addresses to validator info."""
    mapping = {}
    for operator_address, info in validators.items():
        account_address = operator_to_account(operator_address)
        if account_address:
            mapping[account_address] = info
    return mapping


def notify_vote(vote, proposal_id, validator_by_account, slack_webhook):
    """Post vote notification to Slack."""
    voter = vote["voter"]
    options = vote.get("options", [])

    if options:
        option = format_vote_option(options[0]["option"])
    else:
        option = "Unknown"

    timestamp = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M UTC")

    # Find validator info
    validator_info = validator_by_account.get(voter)

    if validator_info:
        moniker = validator_info["moniker"]
        vp = validator_info["voting_power_pct"]
        message = f"{moniker} ({vp:.1f}% VP) votes {option} on proposal {proposal_id} at {timestamp}"
    else:
        # Not a validator vote, or couldn't find mapping
        short_voter = f"{voter[:12]}...{voter[-6:]}" if len(voter) > 20 else voter
        message = f"{short_voter} votes {option} on proposal {proposal_id} at {timestamp}"

    response = requests.post(slack_webhook, json={"text": message}, timeout=10)
    response.raise_for_status()


def format_vote_option(option):
    """Convert vote option enum to human readable."""
    mapping = {
        "VOTE_OPTION_YES": "Yes",
        "VOTE_OPTION_NO": "No",
        "VOTE_OPTION_ABSTAIN": "Abstain",
        "VOTE_OPTION_NO_WITH_VETO": "NoWithVeto",
    }
    return mapping.get(option, option)


def post_error_to_slack(webhook, error_message):
    """Post error to error notification channel."""
    try:
        requests.post(webhook, json={
            "text": f":warning: Gov notification bot error: {error_message}"
        }, timeout=10)
    except Exception:
        # Don't raise if error notification fails
        pass


# For local testing
if __name__ == "__main__":
    import sys

    # Check if environment variables are set
    if not os.environ.get("COSMOS_LCD_URL"):
        os.environ["COSMOS_LCD_URL"] = "https://cosmos-lcd.publicnode.com"

    if not os.environ.get("PROPOSAL_IDS"):
        print("Please set PROPOSAL_IDS environment variable (e.g., export PROPOSAL_IDS=1022)")
        sys.exit(1)

    dry_run = "--dry-run" in sys.argv

    if dry_run:
        print("Running in dry-run mode (no Slack notifications)")

        # Test fetching data
        lcd_url = os.environ["COSMOS_LCD_URL"]
        proposal_ids = os.environ["PROPOSAL_IDS"].split(",")

        for proposal_id in proposal_ids:
            proposal_id = proposal_id.strip()
            if not proposal_id:
                continue

            print(f"\n=== Proposal {proposal_id} ===")

            try:
                votes = fetch_all_votes(lcd_url, proposal_id)
                print(f"Total votes: {len(votes)}")

                validators = fetch_validators(lcd_url)
                print(f"Total bonded validators: {len(validators)}")

                validator_by_account = build_account_to_validator_mapping(validators)

                # Show first few votes
                print("\nSample votes:")
                for i, vote in enumerate(votes[:5]):
                    voter = vote["voter"]
                    options = vote.get("options", [])
                    option = format_vote_option(options[0]["option"]) if options else "Unknown"

                    validator_info = validator_by_account.get(voter)
                    if validator_info:
                        print(f"  {validator_info['moniker']} ({validator_info['voting_power_pct']:.2f}% VP): {option}")
                    else:
                        short_voter = f"{voter[:12]}...{voter[-6:]}" if len(voter) > 20 else voter
                        print(f"  {short_voter}: {option}")

                if len(votes) > 5:
                    print(f"  ... and {len(votes) - 5} more votes")

            except Exception as e:
                print(f"Error processing proposal {proposal_id}: {e}")
    else:
        if not os.environ.get("SLACK_WEBHOOK_URL"):
            print("Please set SLACK_WEBHOOK_URL environment variable for live mode")
            print("Or use --dry-run for testing without Slack")
            sys.exit(1)

        result = handler({}, {})
        print(f"Result: {result}")
